src/other/index.jl

```diff
@@ -108,19 +108,49 @@ function Base.insert!(x::Index, idx::Integer, nm::Symbol)
     x
 end
 
-Base.getindex(x::Index, idx::Symbol) = x.lookup[idx]
-Base.getindex(x::AbstractIndex, idx::Real) = Int(idx)
-Base.getindex(x::AbstractIndex, idx::AbstractVector{Union{Bool, Missing}}) =
-    getindex(x, collect(Missings.replace(idx, false)))
+Base.getindex(x::AbstractIndex, idx::Symbol) = x.lookup[idx]
+Base.getindex(x::AbstractIndex, idx::AbstractVector{Symbol}) = [x.lookup[i] for i in idx]
+
+function Base.getindex(x::AbstractIndex, idx::Real)
+    idx isa Bool && Base.depwarn("Indexing with Bool values is deprecated", :getindex)
+    idx isa Integer || Base.depwarn("Indexing with values that are not Integer is deprecated", :getindex)
+    Int(idx)
+end
+
+function Base.getindex(x::AbstractIndex, idx::AbstractVector{<:Real})
+    any(v -> v isa Bool, idx) && Base.depwarn("Indexing with Bool values is deprecated")
+    all(v -> v isa Integer, idx) || Base.depwarn("Indexing with values that are not Integer is deprecated", :getindex)
+    Vector{Int}(idx)
+end
+
 function Base.getindex(x::AbstractIndex, idx::AbstractVector{Bool})
     length(x) == length(idx) || throw(BoundsError(x, idx))
     find(idx)
 end
-Base.getindex(x::AbstractIndex, idx::AbstractVector{T}) where {T >: Missing} =
-    getindex(x, collect(skipmissing(idx)))
-Base.getindex(x::AbstractIndex, idx::AbstractRange) = [idx;]
-Base.getindex(x::AbstractIndex, idx::AbstractVector{T}) where {T <: Real} = convert(Vector{Int}, idx)
-Base.getindex(x::AbstractIndex, idx::AbstractVector{Symbol}) = [x.lookup[i] for i in idx]
+
+Base.getindex(x::AbstractIndex, idx::AbstractRange) = getindex(x, collect(idx))
+
+# catch all method handling cases when type of idx is not narrowest possible, Any in particular
+# also it handles passing missing values in idx
+function Base.getindex(x::AbstractIndex, idx::AbstractVector)
+    idxs = filter(!ismissing, idx)
+    if length(idxs) != length(idx)
+        Base.depwarn("using missing in column indexing is deprecated", :getindex)
+    end
+    length(idxs) == 0 && return Int[] # special case of empty idxs
+    # below is not very efficient, but has to be done as in Bool case we replaced missing by false
+    # after deprecation period when Missing is disallowed this will not be a problem
+    if idxs[1] isa Bool
+        if eltype(idx) == Union{Bool, Missings.Missing}
+            return getindex(x, Bool[ismissing(v) ? false : v for v in idx])
+        end
+        throw(ArgumentError("Indexing with Bool values is not allowed"))
+    end
+    idxs[1] isa Real && return getindex(x, Vector{Real}(idxs))
+    idxs[1] isa Symbol && return [x.lookup[i] for i in idxs]
```

Comment:
Why not call `getindex` here too for simplicity and consistency?

Reply:
OK, but this will introduce a bit of performance overhead.
I suppose it is not a huge problem as method with incorrect type of `Vector` should not be a typical scenario so I will change it.

Reply:
Are you sure? With inlining, shouldn't be code be identical?

Reply:
The performance overhead is the following:
```
julia> x = Any[Symbol(s) for s in 'A':'z'];

julia> @benchmark convert(Vector{Symbol}, $x)
BenchmarkTools.Trial:
  memory estimate:  544 bytes
  allocs estimate:  1
  --------------
  minimum time:     3.382 μs (0.00% GC)
  median time:      3.616 μs (0.00% GC)
  mean time:        3.750 μs (0.43% GC)
  maximum time:     93.423 μs (89.08% GC)
  --------------
  samples:          10000
  evals/sample:     8
```
by using `convert` (or what I have written) - a new vector will be allocated (checked on 0.6.2 an 0.7) - and I need to convert to make the dispatch work.